compiler/rustc_hir_analysis/src/coherence/orphan.rs

```diff
@@ -99,50 +99,115 @@ fn do_orphan_check_impl<'tcx>(
 
-        let msg = match opt_self_def_id {
-            // We only want to permit nominal types, but not *all* nominal types.
-            // They must be local to the current crate, so that people
-            // can't do `unsafe impl Send for Rc<SomethingLocal>` or
-            // `impl !Send for Box<SomethingLocalAndSend>`.
-            Some(self_def_id) => {
-                if self_def_id.is_local() {
-                    None
-                } else {
-                    Some((
-                        format!(
-                            "cross-crate traits with a default impl, like `{}`, \
-                                    can only be implemented for a struct/enum type \
-                                    defined in the current crate",
-                            tcx.def_path_str(trait_def_id)
-                        ),
-                        "can't implement cross-crate trait for type in another crate",
-                    ))
+        if trait_def_id.is_local() {
+            // If the auto-trait is local, almost anything goes.
+            //
+            //     impl MyAuto for Rc<Something> {}  // okay
+            //     impl<T> !MyAuto for *const T {}   // okay
+            //     impl<T> MyAuto for T {}           // okay
+            //
+            // But there is one important exception: implementing for a trait
+            // object is not allowed.
+            //
+            //     impl MyAuto for dyn Trait {}      // NOT OKAY
+            //     impl<T: ?Sized> MyAuto for T {}   // NOT OKAY
+            //
+            // With this restriction, it's guaranteed that an auto-trait is
+            // implemented for a trait object if and only if the auto-trait is
+            // one of the trait object's trait bounds (or a supertrait of a
+            // bound). In other words `dyn Trait + AutoTrait` always implements
+            // AutoTrait, while `dyn Trait` never implements AutoTrait.
+            //
+            // This is necessary in order for autotrait bounds on methods of
+            // trait objects to be sound.
+            //
+            //     auto trait AutoTrait {}
+            //
+            //     trait ObjectSafeTrait {
+            //         fn f(&self) where Self: AutoTrait;
+            //     }
+            //
+            // We can allow f to be called on `dyn ObjectSafeTrait + AutoTrait`.
+            //
+            // If we didn't deny `impl AutoTrait for dyn Trait`, it would be
+            // unsound for the ObjectSafeTrait shown above to be object safe
+            // because someone could take some type implementing ObjectSafeTrait
+            // but not AutoTrait, unsize it to `dyn ObjectSafeTrait`, and call
+            // .f() which has no concrete implementation (issue #50781).
+            let problematic_kind = match self_ty_kind {
+                ty::Dynamic(..) => Some("trait object"),
+                ty::Param(..) if !self_ty.is_sized(tcx, tcx.param_env(def_id)) => {
```

Comment:
you can also reference trait objects by using projections and opaque types.

You can have impls like
```rust
trait Id { type This: ?Sized; }
impl<T: ?Sized> Id for T {
    type This = T;
}
impl<T: ?Sized> AutoTrait for <T as Id>::This {}
``` 
this impl hits the error "the type parameter `T` is not constrained by the impl trait, self type, or predicates" though so it doesn't matter too much.[^1] Considering that this constraint is necessary for soundness, please make this an exhaustive match.

While I would personally like to merge the two `if trait_def_id.is_local()` branches and only have 1 match on `self_ty.kind()` that's just my personal preference and is fine as is.

[^1]: I can imagine us lifting that requirement for special associated types though, which is why denying it here as well would be good.

Reply:
Fixed, and changed to 1 exhaustive match of TyKind.